### PR TITLE
Fix causal mask in llama for long seq_length

### PR DESCRIPTION
**Pull request by YLGH**
# What does this PR do?

Fix a bug in Llama causal mask creation for when input length > causal_mask.shape (which comes from max_position_embeddings). 

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@ArthurZucker @younesbelkada

### CODE REVIEW

**Comment by YLGH** [comment]
> Will this PR fix #29252 ? (Yi has same architecture as Llama2)

Yeah I think so, (it uses LlamaForCausalLM)

**Comment by YLGH** [comment]
I'm not sure why unit tests are failing, they seem unrelated to my change...

File: src/transformers/models/llama/modeling_llama.py [hunk 1]
```diff
@@ -1061,9 +1061,10 @@ def _update_causal_mask(self, attention_mask, input_tensor):
         device = input_tensor.device
 
         # support going beyond cached `max_position_embedding`
-        if seq_length > self.causal_mask.shape[-1]:
-            causal_mask = torch.full((2 * self.causal_mask.shape[-1], 2 * self.causal_mask.shape[-1]), fill_value=1)
-            self.register_buffer("causal_mask", torch.triu(causal_mask, diagonal=1), persistent=False)
+        causal_mask = self.causal_mask
+        while seq_length > causal_mask.shape[-1]:
+            causal_mask = torch.full((2 * causal_mask.shape[-1], 2 * causal_mask.shape[-1]), fill_value=1)
```
Comment (ArthurZucker) [inline src/transformers/models/llama/modeling_llama.py:1066]:
```suggestion
        while seq_length > causal_mask.shape[-1]:
            new_max_positions = round(seq_length / causal_mask.shape[-1]) * causal_mask.shape[-1]
            causal_mask = torch.full((new_max_positions, new_max_positions), fill_value=1)
```

**Review by ArthurZucker — COMMENTED** [review]
thanks! I don't think a while loop is the best way to achieve this. However we do need a proper fix!

File: src/transformers/models/llama/modeling_llama.py (hunk 1 above)
Comment (ArthurZucker) [reply src/transformers/models/llama/modeling_llama.py:1066]: this should offer a tradeoff between loading too big of a mask and increasing it based on the length of the input

**Comment by BlackSamorez** [comment]
It would also allow one to run `miqu-1-70b` on an RTX 3090 in 2 bits. The model itself is ~19Gb but those masks are too large (around a Gb each and copied sometimes) at `32k` max context length, preventing one from running the model.

**Comment by ArthurZucker** [comment]
I'll prepare some com, but basically you should update the `max_position_embedding` to the appropriate value with which you intend to run the model. This will only affect the causal mask and is no longer taken into account for the ROPE embeddings.

**Comment by ArthurZucker** [comment]
fixed by #29753 !
